### DISCUSSION TITLE
What is the equivalent of Peg.js labels ?

### DISCUSSION THREAD

**Discussion by loloof64** [petitparser/dart-petitparser#134] [Q&A] [answered]
Hi ! What is the equivalent of peg.js labels ?

Let's say for example I want to convert the following Peg.js rule 
```
additive
  = left:multiplicative "+" right:additive { return left + right; }
  / multiplicative
```

This rule uses two labels in the grammar : `left` and `right`. Then it uses them in the parser (`return left+right;`).

What is the best way to do it in PetitParser ?

I would like to use such a feature because in many rules I've got optional parts, and so it would make my task easier.

Also I'm using the same structure as in the examples (prolog, json)
* a grammar file
* a parser file
* later I plan to set up an evaluator file

Regards

**Comment by renggli** [comment]
PetitParser doesn't use code generation (like peg.js). This would enable things like labels, but also comes at the cost of introducing another language and making it impossible to arbitrarily mix Dart and parser code. This might change hopefully soon when Dart gets support for [static metaprogramming](https://github.com/dart-lang/language/issues/1482).

In the meantime a sequence of parsers generates a list of parse results that you can access by index, i.e.

```dart
Parser additive() => (ref0(multiplicative) & char("+") & ref0(additive)).map((values) => values[0] + values[2])
     / ref(multiplicative);
```

If this is a common pattern, I recommend to extract it to a helper method, i.e.

```dart
Parser term(Parser left, Parser op, Parser right, int Function(int left, int right) callback) {
  return (left & op & right).map((values) => callback(values[0], values[2])) / ref(left);
}

Parser additive() => term(ref0(multiplicative), char("+"), ref0(additive), (left, right) => values[0] + values[2]);
```

Also note that for expressions there is [an expression-builder](https://github.com/petitparser/dart-petitparser/blob/443cce91558dbb8b608032c9eaa43f6e68cf106c/lib/src/expression/builder.dart#L46) that does exactly this out of the box:

```dart
...
builder.group()
   ..left(char('+').trim(), (a, op, b) => a + b)
...
```

**Reply by loloof64** [reply]
Thank you very much for your answer.

In fact, in the grammar I'm writing, I've written the following rule :

```dart
import 'package:petitparser/petitparser.dart';

class PgnGrammarDefinition extends GrammarDefinition {

    Parser game() =>
      ref0(tags).optional() & ref0(comments).optional() & ref0(pgn);

}
```

And in the parser file, I've written something like this
```dart
import 'package:petitparser/petitparser.dart';
import 'grammar.dart';

class PgnParserDefinition extends PgnGrammarDefinition {
  var _messages = [];

   @override
   Parser game() => super.game().map((each) {
      var message = _messages; _messages = [];
      return {
        'tags': each[0],
        'gameComment': each[1],
        'moves': each[2],
        'messages': message,
     };
  });
}
```

So is it warranted that in the overriden `game()` function:
* each[0] matches the `ref0(tags)` even if not providen as it is optionnal ?
* each[1] matches the `ref0(comments)` even if not providen ?
* each[2] matches the `ref0(pgn)` ?
So that there is no need to make use condtionnal, which for sure I'll code this in an ugly way ?

Regards

**Comment by renggli** [comment, marked as answer]
Yes, that's correct. The optional parsers get a fixed entry in the resulting list. By default the resulting absent value will be `null`, but there is also the option to specify any other default value, see the [documentation](https://pub.dev/documentation/petitparser/latest/parser/OptionalParserExtension.html).

**Reply by loloof64** [reply]
Thank you very much 😃
